migrations/migration.py

```diff
@@ -0,0 +1,264 @@
+            audit_entry(
+                message,
+                level='info'
+            )
+
+            raise KeyError(message)
+
+        applied_migrations = None
+        unapplied_migrations = None
+        if direction == "upgrade":
+            unapplied_migrations = self.get_unapplied_migrations(current_migration)
+        elif direction == "downgrade" and current_migration is not None:
+            applied_migrations = self.get_previous_migration(current_migration)
+            unapplied_migrations = current_migration
+
+        if not unapplied_migrations or unapplied_migrations == 'None':
+            raise ValueError("No valid migration files to run.")
+
+        if direction == "upgrade":
+            # Run all available migrations
+            for migration in unapplied_migrations:
+                self.run_migration(direction, migration, migration)
+        if direction == "downgrade":
+            # Run one migration down
+            self.run_migration(direction, unapplied_migrations, applied_migrations)
+
+    def run_migration(self, direction: str, next_migration: str, applied_migration: str):
+        """Run single migration based on the specified direction ("upgrade" or "downgrade")."""
+        # Update the migration to acquire most recent updates in the system
+        migration_path = os.path.join(self.migrations_dir, self.migrations_locations[next_migration] + ".py")
+        try:
+            audit_entry(
+                "running migration",
+                level='info'
+            )
+
+            migration_module = imp.load_source('migration_module', migration_path)
+
+            if direction == "upgrade":
+                migration_module.upgrade()
+            elif direction == "downgrade":
+                migration_module.downgrade()
+
+            self.update_latest_applied_migration_in_fhir(applied_migration)
+        except Exception as e:
+            message = f"Error executing migration {applied_migration}: {e}"
+            audit_entry(
+                message,
+                level='debug'
+            )
+
+    def get_unapplied_migrations(self, applied_migration) -> list:
+        """Retrieve all migrations after the applied migration."""
+        return self.migration_sequence.get_sublist(applied_migration)
+
+    def get_previous_migration(self, current_migration) -> str:
+        """Retrieve the previous migration."""
+        return self.migration_sequence.previous(current_migration)
+
+    def get_latest_created_migration(self) -> str:
+        """Retrieve the latest migration in the entire migration sequence."""
+        return self.migration_sequence.head
+
+    ## FHIR MANAGEMENT LOGIC
+    def get_latest_applied_migration_from_fhir(self) -> str:
+        """Retrieve the latest applied migration migration id from FHIR."""
+        # Logic to retrieve the latest applied migration number from FHIR
+        basic = MigrationManager.get_resource()
```

Comment:
i see opportunity to DRY this out a bit.  both this method and the next have a half dozen lines of code that could be done once in `MigrationManager.get_resource()`